### PR TITLE
[FIX 7078] better error messages

### PR DESCRIPTION
**Pull request by alanjds**
Better handling of errors, and more informative error messages. Most of them include the last day that an ingestion worked, and they all mention that failed data will be stored to be retried later.

### CODE REVIEW

File: src/pages/Background/index.js [hunk 1]
```diff
@@ -137,33 +139,169 @@ class PpdnsBackground {
...
+
+    let snoozedUntil = (await this.storage.get(SETTINGS_KEY))[SETTINGS_KEY].snoozedUntil;
+    if (Number(snoozedUntil) >= Date.now()){
+      // Snoozed.
+      console.info('Notification: %s [snoozed until %s]', errorname, snoozedUntil);
+      return
+    }
+
+    let message = ('The data will be held waiting the API Key to be checked from your account settings.'
+    + '\n\xa0\nClick here to open polyswarm.network/account/api-keys in a new tab');
```
Comment (alanjds) [inline src/pages/Background/index.js:199]: Native english speakers, please double check this piece.

File: src/pages/Background/index.js [hunk 2]
```diff
@@ -137,33 +139,169 @@ class PpdnsBackground {
...
+      return
+    }
+
+    let lastSuccess = await this.getLastSuccessString();
+    let notificationOptions = {
+      type: 'basic',
+      iconUrl: 'icon-34.png',
+      title: 'We had a problem (not in Houston)',
+      message: ('The data will be held and retransmitted as soon as we fix the Polyswarm side.'
+                + '\nLast successful transmission was ' + lastSuccess),
```
Comment (alanjds) [inline src/pages/Background/index.js:244]: Native english speakers, please double check this piece. Including the title.

Please send suggestions via Github if possible.

File: src/pages/Background/index.js [hunk 3]
```diff
@@ -137,33 +139,169 @@ class PpdnsBackground {
...
+      return
+    }
+
+    let lastSuccess = await this.getLastSuccessString();
+    let notificationOptions = {
+      type: 'basic',
+      iconUrl: 'icon-34.png',
+      title: 'Connection error submitting data',
+      message: ('The data will be held and retransmitted as soon as a conection became possible.'
+                + '\nLast successful transmission was ' + lastSuccess),
```
Comment (alanjds) [inline src/pages/Background/index.js:283]: Native english speakers, please double check this piece.

File: src/pages/Background/index.js [hunk 4]
```diff
@@ -174,11 +312,19 @@ class PpdnsBackground {
       return
     }
 
+    let message = '';
+    let lastSuccess = await this.getLastSuccessString();
+    if (lastSuccess == 'today'){
+      message = 'Could not send the new data right now, but already sent something today.\nWill keep the data and try again soon.';
+    }else{
+      message = 'Last successful submission was ' + lastSuccess + '.\nWill keep the data and try again soon.';
```
Comment (alanjds) [inline src/pages/Background/index.js:320]: Native english speakers, please double check this piece.

File: src/pages/Background/index.js [hunk 5]
```diff
@@ -197,47 +343,73 @@ class PpdnsBackground {
...
+    await this.snoozeNotifications(snoozedUntil);
+  }
+
+  async getLastSuccessString() {
+    let apiKeyCheckedDate = '0'+(await this.storage.get(SETTINGS_KEY))[SETTINGS_KEY].apiKeyCheckedDate;
+    if (Number(apiKeyCheckedDate) >= Date.now() - 86400000){  // yesterday
+      return 'today'
+    }else{
+      let lastDate = new Date(Number(apiKeyCheckedDate));
+      return 'by ' + lastDate.toLocaleDateString()
```
Comment (alanjds) [inline src/pages/Background/index.js:355]: Native english speakers, please double check this piece.

The idea is to return `today` or `by 2024/12/20` or what makes sense in the browser language settings.

That will turn into `last time it worked was by 2024...` or `last time it worked was today.`
If this is not correct in English, please provide a suggestion for the `by ` thing.

Comment (alanjds) [reply src/pages/Background/index.js:355]: @supernothing & @mrtizmoatwork I am marking you here to be able to check the English spellings.

File: src/pages/Background/index.js (hunk 1 above)
Comment (mjbradford89) [reply src/pages/Background/index.js:199]:
```suggestion
    let message = ('Your telemetry data will be saved locally until you enter an API Key from your account settings.'
    + '\n\xa0\nClick here to open polyswarm.network/account/api-keys in a new tab');
```

File: src/pages/Background/index.js (hunk 2 above)
Comment (mjbradford89) [reply src/pages/Background/index.js:244]:
```suggestion
      title: 'We had an issue submitting telemetry data.',
      message: ('The data will be saved and resubmitted as soon as it is possible.'
                + '\nLast successful transmission was ' + lastSuccess),
```

File: src/pages/Background/index.js [hunk 6]
```diff
@@ -137,33 +139,169 @@ class PpdnsBackground {
...
+      return
+    }
+
+    let lastSuccess = await this.getLastSuccessString();
+    let notificationOptions = {
+      type: 'basic',
+      iconUrl: 'icon-34.png',
+      title: 'Connection error submitting data',
+      message: ('The data will be held and retransmitted as soon as a conection became possible.'
+                + '\nLast successful transmission was ' + lastSuccess),
```
Comment (mjbradford89) [inline src/pages/Background/index.js:283]:
```suggestion
      title: 'There was an issue connecting to the internet.',
      message: ('The data will be saved and resubmitted once the connection is fixed .'
                + '\nLast successful transmission was ' + lastSuccess),
```

File: src/pages/Background/index.js (hunk 4 above)
Comment (mjbradford89) [reply src/pages/Background/index.js:320]:
```suggestion
    if (lastSuccess == 'today'){
      message = 'The data could not be sent right now, but some data has been sent today already.\nIt will be saved and resubmitted again soon.';
    }else{
      message = 'There is an issue submitting telemetry data. Last successful submission was ' + lastSuccess + '.\nOnce the issue is fixed the data will be resubmitted.';
```

File: src/pages/Background/index.js (hunk 5 above)
Comment (mjbradford89) [reply src/pages/Background/index.js:355]:
```suggestion
      return 'on ' + lastDate.toLocaleDateString()
```

I think this sounds right:
>last time it worked was on 2024/12/20

**Review by mjbradford89 — CHANGES_REQUESTED** [review]
Updated with some English suggestions.

Also unrelated to your changes, I noticed the API key validation is still using v2 here: https://github.com/polyswarm/ppdns-chrome-extension/pull/48/files#diff-b6bb0030ac3f7012a3807915b291ab879b577c13cbbb3203f0b35edeb1b54481R95

Would you be able to change that to v3 in this PR please?  I think the v2 api is no longer working.

File: src/pages/Background/index.js (hunk 2 above)
Comment (alanjds) [reply src/pages/Background/index.js]: I updated the message but the title became too big for one line:
Screenshot_20241004_071315

Would the follow be acceptable instead?
Screenshot_20241004_072606

File: src/pages/Background/index.js (hunk 6 above)
Comment (alanjds) [reply src/pages/Background/index.js]: I updated the message but the title got too big for one line:
Screenshot_20241004_070643

Would be acceptable something like this instead?
Screenshot_20241004_073439

**Comment by alanjds** [comment]
@mjbradford89 Thanks for the updates.

Some of the titles became too big for one line. I changed these from "We had an issue submitting" to "Could not submit" and "There is an issue connecting" to "Could not connect".

Could you please take another look?

File: src/pages/Background/index.js (hunk 2 above)
Comment (mjbradford89) [reply src/pages/Background/index.js]: Yep I think that is better actually :+1:

**Comment by mjbradford89** [comment]
hey @alanjds I think these changes look good.  But I am having an issue when testing.  When I install the extension and enter an API key, the request that validates it returns a 500 so the key is not validated correctly.  I think this is probably because the /v3/instance endpoint works differently than v2 :cry:.

image

**Comment by mjbradford89** [comment]
I created a pr that should fix the 500 error in AI: https://github.com/polyswarm/artifact-index/pull/1429/files

Hopefully this will fix the issue with the api key validation and no more changes are needed here.

**Comment by alanjds** [comment]
> I created a pr that should fix the 500 error in AI. Hopefully this will fix the issue with the api key validation and no more changes are needed here.

I will revert the v2 -> v3 change. During my tests, v2 was working good

**Comment by alanjds** [comment]
> hey @alanjds I think these changes look good. But I am having an issue when testing. When I install the extension and enter an API key, the request that validates it returns a 500 so the key is not validated correctly. I think this is probably because the /v3/instance endpoint works differently than v2 😢.
> 
> image

@mjbradford89 I changed back to v2 in a828803. Please take another look.

**Review by mjbradford89 — APPROVED** [review]

**Pull request merged**